src/adaptix/_internal/morphing/generic_provider.py

```diff
@@ -120,38 +192,51 @@ def _provide_loader(self, mediator: Mediator, request: LoaderRequest) -> Loader:
             allowed_values_with_types = self._get_allowed_values_collection(
                 [(type(el), el) for el in cleaned_args]
             )
-            allowed_values_repr = set(cleaned_args)
 
             # since True == 1 and False == 0
             def literal_loader_sc(data):
                 if (type(data), data) in allowed_values_with_types:
                     return data
                 raise BadVariantError(allowed_values_repr, data)
 
-            return literal_loader_sc
+            return self._literal_loader_with_enum(
+                literal_loader_sc, enum_loaders, allowed_values_with_types
+            )
 
         allowed_values = self._get_allowed_values_collection(cleaned_args)
-        allowed_values_repr = set(cleaned_args)
 
         def literal_loader(data):
             if data in allowed_values:
                 return data
             raise BadVariantError(allowed_values_repr, data)
 
-        return literal_loader
+        return self._literal_loader_with_enum(literal_loader, enum_loaders, allowed_values)
 
     def _provide_dumper(self, mediator: Mediator, request: DumperRequest) -> Dumper:
         norm = try_normalize_type(get_type_from_request(request))
         cleaned_args = [strip_annotated(arg) for arg in norm.args]
         enum_cases = [arg for arg in cleaned_args if isinstance(arg, Enum)]
-        if enum_cases:
-            raise CannotProvide(
-                f"Enum inside Literal isn't supported yet, found {enum_cases}",
-                is_terminal=True,
-                is_demonstrative=True
-            )
 
-        return as_is_stub
+        if not enum_cases:
+            return as_is_stub
+
+        enum_dumpers = self._fetch_enum_dumpers(
+            mediator, request, [type(case) for case in enum_cases]
+        )
+
+        def literal_dumper(data):
+            if isinstance(data, Enum):
+                return enum_dumpers[type(data)](data)
+            return data
+
+        enum_dumper = list(enum_dumpers.values())[0]
+
+        def literal_dumper_with_single_enum(data):
+            if isinstance(data, Enum):
+                return enum_dumper(data)
+            return data
+
+        return literal_dumper_with_single_enum if len(enum_cases) == 1 else literal_dumper
```

Comment:
```suggestion
        if len(enum_dumpers) == 1:
            enum_dumper = enum_dumpers[0]
            
            def literal_dumper_with_single_enum(data):
                if isinstance(data, Enum):
                    return enum_dumper(data)
                return data
                
            return literal_dumper_with_single_enum

        def literal_dumper_with_enums(data):
            if isinstance(data, Enum):
                return enum_dumpers[type(data)](data)
            return data
         
        return literal_dumper_with_enums
```

You can avoid creating excess closures